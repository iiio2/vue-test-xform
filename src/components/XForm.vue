<template>
  <form>
    <div v-for="(field, index) in fields" :key="index">
      <component
        :is="field.component ?? 'input'"
        :value="modelValue[field.field]"
        :name="field.field"
        :placeholder="field.placeholder"
        :type="field.type"
        @input="onInput($event)"
      >
        {{ modelValue[field.field] ?? field.placeholder }}
      </component>
    </div>
  </form>
</template>

<script>
export default {
  props: {
    modelValue: {
      type: Object,
      required: true,
    },
    fields: {
      type: Array,
      required: true,
    },
  },
  emits: ["update:modelValue"],
  methods: {
    onInput(event, index) {
      let newModelValue = { ...this.modelValue };
      newModelValue[event.target.name] = event.target.value;
      this.$emit("update:modelValue", newModelValue);
    },
  },
};
</script>
