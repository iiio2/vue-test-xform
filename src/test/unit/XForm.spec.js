import { mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";
import XForm from "../../components/XForm.vue";

describe("XForm", () => {
  it("should be truthy", () => {
    expect(XForm).toBeTruthy();
  });
  it("should accept object in v-model and array in props", () => {
    const wrapper = mount(XForm, {
      props: {
        modelValue: {
          name: "Alex",
          dateOfBirth: "2023-01-06",
          eyeColor: "blue",
        },
        fields: [
          {
            field: "name",
            component: "input",
            type: "text",
            placeholder: "Name",
          },
        ],
      },
    });
    expect(wrapper.props().modelValue).toStrictEqual({
      name: "Alex",
      dateOfBirth: "2023-01-06",
      eyeColor: "blue",
    });
    expect(wrapper.props().fields).toStrictEqual([
      {
        field: "name",
        component: "input",
        type: "text",
        placeholder: "Name",
      },
    ]);
  });
  it("should render each string as separate input", async () => {
    const wrapper = mount(XForm, {
      props: {
        modelValue: {
          name: "Alex",
          dateOfBirth: "2023-01-06",
          eyeColor: "blue",
        },
        fields: [
          {
            field: "name",
            component: "input",
            type: "text",
            placeholder: "Name",
          },
        ],
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.findAll("input").length).toBe(1);
  });

  it("writes back to the array when input value is changed", async () => {
    const wrapper = mount(XForm, {
      props: {
        modelValue: {
          name: "Alex",
          dateOfBirth: "2023-01-06",
          eyeColor: "blue",
        },
        fields: [
          {
            field: "name",
            component: "input",
            type: "text",
            placeholder: "Name",
          },
        ],
      },
    });
    await wrapper.vm.$nextTick();
    const input = wrapper.findAll("input").at(0);
    await input.setValue("new input");
    expect(wrapper.emitted("update:modelValue")[0][0]).toStrictEqual({
      name: "new input",
      dateOfBirth: "2023-01-06",
      eyeColor: "blue",
    });
  });
  ///// #####
  it("example works as expected (compare the snapshot)", async () => {
    const wrapper = mount(XForm, {
      props: {
        modelValue: {
          name: "Alex",
          dateOfBirth: "2023-01-06",
          eyeColor: "blue",
        },
        fields: [
          {
            field: "name",
            component: "input",
            type: "text",
            placeholder: "Name",
          },
        ],
      },
    });
    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
