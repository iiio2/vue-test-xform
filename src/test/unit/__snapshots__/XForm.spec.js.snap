// Vitest Snapshot v1

exports[`XForm > example works as expected (compare the snapshot) 1`] = `
"<form>
  <div><input name=\\"name\\" placeholder=\\"Name\\" type=\\"text\\"></div>
</form>"
`;
